code/controllers/subsystem/fluids.dm

```diff
@@ -0,0 +1,266 @@
+	flags = SS_BACKGROUND|SS_KEEP_TIMING
+	runlevels = RUNLEVEL_GAME|RUNLEVEL_POSTGAME
+
+	/// What this subsystem processes. Yes, it's called 'case' entirely because of 'switch() case: '.
+	var/case = NONE
+
+	// Fluid spread processing:
+	/// The amount of time before a fluid node is created and when it spreads.
+	var/spread_wait = 1 SECONDS
+	/// The number of buckets in the spread carousel.
+	var/tmp/num_spread_buckets
+	/// The set of buckets containing fluid nodes to spread.
+	var/list/spread_carousel
+	/// The index of the spread carousel bucket currently being processed.
+	var/spread_bucket_index
+	/// The set of fluid nodes we are currently processing spreading for.
+	var/list/currently_spreading
+	/// Whether the subsystem has resumed spreading fluid.
+	var/resumed_spreading
+
+	// Fluid effect processing:
+	/// The amount of time between effect processing ticks for each fluid node.
+	var/effect_wait = 1 SECONDS
+	/// The number of buckets in the effect carousel.
+	var/num_effect_buckets
+	/// The set of buckets containing fluid nodes to process effects for.
+	var/list/effect_carousel
+	/// The index of the currently processing bucket on the effect carousel.
+	var/effect_bucket_index
+	/// The set of fluid nodes we are currently processing effects for.
+	var/list/currently_processing
+	/// Whether the subsystem has resumed processing fluid effects.
+	var/resumed_effect_processing
+
+/datum/controller/subsystem/fluids/Initialize(start_timeofday)
+	initialize_waits()
+	if (case & SS_PROCESSES_SPREADING)
+		initialize_spread_handling()
+	if (case & SS_PROCESSES_EFFECTS)
+		initialize_effect_handling()
+	return ..()
+
+/**
+ * Initializes the subsystem waits.
+ *
+ * Ensures that the subsystem's fire wait evenly splits the spread and effect waits.
+ */
+/datum/controller/subsystem/fluids/proc/initialize_waits()
+	case = NONE
+	if (spread_wait > 0)
+		case |= SS_PROCESSES_SPREADING
+	if (effect_wait > 0)
+		case |= SS_PROCESSES_EFFECTS
+
+	switch (case)
+		if (SS_PROCESSES_SPREADING|SS_PROCESSES_EFFECTS)
+			wait = Gcd(wait, Gcd(spread_wait, effect_wait))
+		if (SS_PROCESSES_SPREADING)
+			if (spread_wait > wait)
+				spread_wait = wait * round(spread_wait / wait, 1)
+			else
+				wait = spread_wait
+			effect_wait = wait
+		if (SS_PROCESSES_EFFECTS)
+			if (effect_wait > wait)
+				effect_wait = wait * round(effect_wait / wait, 1)
+			else
+				wait = effect_wait
+			spread_wait = wait // It needs to be _something_ to prevent runtimes.
+			stack_trace("[src] is a fluid subsystem that does not process fluid spread. Just make it a regular subsystem please.")
+		else
+			spread_wait = wait
+			flags |= SS_NO_FIRE
+			stack_trace("[src] is a fluid subsystem without any purpose. Cull it soon please.")
+
+/**
+ * Initializes the carousel used to process fluid spreading.
+ */
+/datum/controller/subsystem/fluids/proc/initialize_spread_handling()
+	num_spread_buckets = round(spread_wait / wait)
+	spread_wait = wait * num_spread_buckets
```

Comment:
why this line?

Reply:
It's to make completely certain that the delta_time associated with the fluid spread is in sync with the amount of time between processes.